django-cloudlaunch/cloudlaunch/tests.py

```diff
@@ -140,3 +162,258 @@ def test_create_cloud_creds(self):
+        # check that ApplicationDeploymentTask was created, will throw
+        # DoesNotExist if missing
+        task = ApplicationDeploymentTask.objects.get(action='HEALTH_CHECK',
+                                                     celery_id=celery_id,
+                                                     deployment=app_deployment)
+        self.assertIsNotNone(task)
+
+    @patch("cloudlaunch.tasks.restart_appliance.delay")
+    def test_create_restart_task(self, restart_task):
+        """Test creating a RESTART type task."""
+        app_deployment = self._create_test_deployment(user=self.user)
+        url = reverse('deployment_task-list',
+                      kwargs={'deployment_pk': app_deployment.id})
+        celery_id = str(uuid.uuid4())
+        restart_task.return_value = AsyncResult(celery_id)
+        response = self.client.post(url, {'action': 'RESTART'})
+        self.assertEqual(response.status_code, 201)
+        restart_task.assert_called_with(
+            app_deployment.id,
+            app_deployment.credentials.as_dict())
+        self.assertEqual(response.data['celery_id'], celery_id)
+        self.assertEqual(response.data['action'], 'RESTART')
+        # check that ApplicationDeploymentTask was created, will throw
+        # DoesNotExist if missing
+        task = ApplicationDeploymentTask.objects.get(action='RESTART',
+                                                     celery_id=celery_id,
+                                                     deployment=app_deployment)
+        self.assertIsNotNone(task)
+
+    @patch("cloudlaunch.tasks.delete_appliance.delay")
+    def test_create_delete_task(self, delete_task):
+        """Test creating a DELETE type task."""
+        app_deployment = self._create_test_deployment(user=self.user)
+        url = reverse('deployment_task-list',
+                      kwargs={'deployment_pk': app_deployment.id})
+        celery_id = str(uuid.uuid4())
+        delete_task.return_value = AsyncResult(celery_id)
+        response = self.client.post(url, {'action': 'DELETE'})
+        self.assertEqual(response.status_code, 201)
+        delete_task.assert_called_with(
+            app_deployment.id,
+            app_deployment.credentials.as_dict())
+        self.assertEqual(response.data['celery_id'], celery_id)
+        self.assertEqual(response.data['action'], 'DELETE')
+        # check that ApplicationDeploymentTask was created, will throw
+        # DoesNotExist if missing
+        task = ApplicationDeploymentTask.objects.get(action='DELETE',
+                                                     celery_id=celery_id,
+                                                     deployment=app_deployment)
+        self.assertIsNotNone(task)
+
+    def test_only_one_launch_task(self):
+        """Test LAUNCH task not allowed if one already exists."""
+        app_deployment = self._create_test_deployment(user=self.user)
+        ApplicationDeploymentTask.objects.create(
```

Comment:
I think this test has really highlighted a design issue with the existing cloudlaunch code - we should probably move the duplication launch check from the serializer to the model?

Reply:
Yeah I think that's a good idea.

Reply:
@nuwang after doing some additional reading, the way I understand it in DRF you should do validation in the serializer, not in the model.  See [Differences between ModelSerializer validation and ModelForm|http://www.django-rest-framework.org/topics/3.0-announcement/#differences-between-modelserializer-validation-and-modelform]

> This change also means that we no longer use the .full_clean() method on model instances, but instead perform all validation explicitly on the serializer. This gives a cleaner separation, and ensures that there's no automatic validation behavior on ModelSerializer classes that can't also be easily replicated on regular Serializer classes.
> ...
> The one difference that you do need to note is that the .clean() method will not be called as part of serializer validation, as it would be if using a ModelForm. Use the serializer .validate() method to perform a final validation step on incoming data where required.

`.clean()` is the most natural place to put the model validation but DRF recommends against that. The way it is currently implemented, with the `validate_action()` method, is the recommended way.

Reply:
I think the original reason this would have been useful was to support the internal API layer. In that case, when attempting to save a deployment task, the duplicate task checker logic wouldn't have run, since it's in the serializer. I still wonder whether this constraint should ideally be double checked in the save() method, since the constraint shouldn't be violated no matter how we save it?